server/gfmd/auth_uid.c

```diff
@@ -255,10 +261,35 @@ auth_uid_to_global_username_kerberos(void *closure,
 	const char *auth_user_id, enum gfarm_auth_id_role *auth_user_id_rolep,
 	char **global_usernamep)
 {
-	return (auth_uid_to_global_username_server_auth(closure,
-	  auth_user_id, gfarm_kerberos_principal_get_hostname,
-	    auth_user_id_rolep, global_usernamep,
-	    "auth_uid_to_global_username_kerberos"));
+	enum gfarm_auth_id_role auth_user_id_role = *auth_user_id_rolep;
```

Comment:
this auth_uid_to_global_username_kerberos() implemetation has the following problems:

* This function should convert a kerberos principal name (e.g. soda@kerberos.domain.name) in auth_user_id to  a gfarm global username in *global_user_namep.  But this implementation converts a gfarm global username in auth_user_id to a kerberos principal name which is returned by user_auth_id() instead.  In other words, the conversion is performed in the opposite direction than it should be.
* This implementation breaks the authentication results when auth_user_id_role is GFARM_AUTH_ID_ROLE_SPOOL_HOST or GFARM_AUTH_ID_ROLE_METADATA_HOST

What you should actually do is to add struct user *(*user_lookup_by_auth_user_id)(const char *auth_user_id) to the input parameter of auth_uid_to_global_username_server_auth() and pass user_lookup_by_kerberos_principal function as this parameter in the Kerberos case.
Pass user_lookup_gsi_dn as the parameter for the gsi case and the tls_client_certificate case, as it is now.

Reply:
commit [430e073](https://github.com/oss-tsukuba/gfarm/pull/1147/commits/430e073272ae3bc3cba42baa80c86b579c7f5ee8) does not solve this issue, and still breaks the GFARM_AUTH_ID_ROLE_SPOOL_HOST and GFARM_AUTH_ID_ROLE_METADATA_HOST case.
Please fix it by making auth_uid_to_global_username_kerberos() call auth_uid_to_global_username_server_auth() as original implmentation and previous comment.